experiments/storm/src/main/scala/io/gearpump/experiments/storm/processor/StormProcessor.scala

```diff
@@ -19,96 +19,89 @@
-      val freq = stormConfig.get(TICK_TUPLE_FREQ_SECS).asInstanceOf[java.lang.Long]
-      val values = new JArrayList[Object](1)
-      values.add(freq)
-      val tickTuple = new TupleImpl(getTickTupleContext, values, SYSTEM_TASK_ID, SYSTEM_STREAM_ID, null)
-      scheduler = taskContext.schedule(new FiniteDuration(freq, TimeUnit.SECONDS),
-        new FiniteDuration(freq, TimeUnit.SECONDS))({
-        self ! tickTuple
-      })
-    }
+    implicit val timeout = Constants.FUTURE_TIMEOUT
+    val dagFuture = askAppMasterForDAG(taskContext.appMaster)
+    val dag = Await.result(dagFuture, timeout.duration)
+
+    val taskToComponent = getTaskToComponent(dag)
+    topologyContextBuilder = new TopologyContextBuilder(topology, stormConfig, taskToComponent)
+    val topologyContext = topologyContextBuilder.build(stormTaskId)
+
+    taskToTopologyContext += stormTaskId -> topologyContext
 
+    outputCollector = new StormOutputCollector(taskContext, topologyContext)
     val delegate = new StormBoltOutputCollector(outputCollector)
-    val topologyContext = topologyContextBuilder.buildContext(pid, boltId)
     bolt.prepare(stormConfig, topologyContext, new OutputCollector(delegate))
+
+    setupTickTuple(topologyContext)
   }
 
   override def onNext(msg: Message): Unit = {
     val stormTuple = msg.msg.asInstanceOf[StormTuple]
     outputCollector.setTimestamp(msg.timestamp)
-    bolt.execute(stormTuple.toTuple(topologyContextBuilder))
+    val sourceId = stormTuple.stormTaskId
+    val topologyContext = taskToTopologyContext.getOrElse(sourceId, topologyContextBuilder.build(sourceId))
+    val tuple = stormTuple.toTuple(topologyContext)
+    bolt.execute(tuple)
   }
 
   override def onStop(): Unit = {
     if (scheduler != null) {
       scheduler.cancel()
     }
+    scheduler = null
   }
 
-  override def receiveUnManagedMessage: Receive = {
-    case tickTuple: TupleImpl =>
-      bolt.execute(tickTuple)
-    case msg =>
-      LOG.error("Failed! Received unknown message " + "taskId: " + taskContext.taskId + ", " + msg.toString)
-  }
-
-  private def getTickTupleContext: GeneralTopologyContext = {
-    val taskToComponent = new JHashMap[Integer, String]
-    taskToComponent.put(SYSTEM_TASK_ID, SYSTEM_COMPONENT_ID)
-    val componentToSortedTasks = new JHashMap[String, JList[Integer]]
-    val tasks = new JArrayList[Integer](1)
-    tasks.add(SYSTEM_TASK_ID)
-    componentToSortedTasks.put(SYSTEM_COMPONENT_ID, tasks)
-    val streamToFields = new JHashMap[String, Fields]
-    streamToFields.put(SYSTEM_STREAM_ID, new Fields(SYSTEM_COMPONENT_OUTPUT_FIELDS))
-    val componentToStreamToFields = new JHashMap[String, JMap[String, Fields]]
-    componentToStreamToFields.put(SYSTEM_COMPONENT_ID, streamToFields)
-    new GeneralTopologyContext(topology, stormConfig, taskToComponent, componentToSortedTasks,
-      componentToStreamToFields, null)
+  private def setupTickTuple(topologyContext: TopologyContext): Unit = {
```

Comment:
Maybe we can create a constant mesasge Tick somewhere.

val Tick = 

And changing this function to something like scheduleTick
And for flow control purpose, maybe we should use scheduleOnce instead.